ipykernel/kernelbase.py

```diff
@@ -396,28 +403,33 @@ async def dispatch_shell(self, msg):
             self.log.warning("Unknown message type: %r", msg_type)
         else:
             self.log.debug("%s: %s", msg_type, msg)
+            asyncio.run_coroutine_threadsafe(self.call_handler(shell_id, handler, idents, msg), self.shell_threads[shell_id].io_loop.asyncio_loop)
```

Comment:
This cannot work because you will send replies on the same socket from two different threads, and ZMQ sockets are NOT thread safe. Same issue if you authorize stdin in the subshell.

Reply:
Yes, but this can be solved by adding locks when sending messages.

Reply:
The whole point of ZeroMQ is to avoid to share sockets (or anything actually) between threads, do not use multithreading primitives to synchronize threads, they should be synchronized via interprocess sockets. Even with a lock, you have no guarantee that the socket will handle both threads correclty.

Reply:
If that is true (I could not find any documentation confirming what you said), then it means that we should use a new ZMQ socket per subshell.

Reply:
Thanks, that is quite clear indeed :smile: 
"To make utterly perfect MT programs (and I mean that literally), we don’t need mutexes, locks, or any other form of inter-thread communication except messages sent across ZeroMQ sockets."
So, do we agree that we should let down the approach consisting of one ZMQ channel for all subshells, and instead have one ZMQ channel per subshell?

Reply:
Indeed I find the second approach cleaner since we delegate the routing to ZeroMQ (although it requires additional changes in the server). However, for the sake of completeness (and if you want to still experiment with the first approach), the way to implement it would be:
- one thread polling the shell sockets and additional PAIR sockets for interthread communication. 
- one thread for the main shell and one per subshell.

When a message arrives on the shel, the polling thread parses it and sends it to the corresponding thread via the pair socket. The executing thread should push the answer on that pair socket instead of sending it on the shell socket.

Reply:
Yes I'm currently implementing it but:

> one thread polling the shell sockets and additional PAIR sockets for interthread communication.

There is only one shell socket and I'm using queues for inter-thread communication instead of sockets.

> one thread for the main shell and one per subshell.

The main shell is in the main thread, each subshell has its own thread.